source/server/api_listener_impl.cc

```diff
@@ -31,6 +31,33 @@ void ApiListenerImplBase::SyntheticReadCallbacks::SyntheticConnection::raiseConn
   }
 }
 
+// This class wraps a HttpConnectionManager used as an Http::ApiListener and the associated
+// SyntheticReadCallbacks to ensure that both objects have the same lifetime.
+class HttpApiListener::HttpConnectionManagerWrapper : public Http::ApiListener {
+public:
+  HttpConnectionManagerWrapper(HttpApiListener& parent, Event::Dispatcher& dispatcher)
+      : read_callbacks_(parent, dispatcher),
+        http_connection_manager_(parent.http_connection_manager_factory_(read_callbacks_)) {}
+
+  ~HttpConnectionManagerWrapper() override {
+    // The Http::ConnectionManagerImpl is a callback target for the read_callback_.connection_. By
+    // raising connection closure, Http::ConnectionManagerImpl::onEvent is fired. In that case the
+    // Http::ConnectionManagerImpl will reset any ActiveStreams it has.
+    read_callbacks_.connection_.raiseConnectionEvent(Network::ConnectionEvent::RemoteClose);
+  }
+
+  Http::RequestDecoder& newStream(Http::ResponseEncoder& response_encoder,
+                                  bool is_internally_created = false) override {
+    return http_connection_manager_->newStream(response_encoder, is_internally_created);
+  }
+
+  SyntheticReadCallbacks& readCallbacksForTest() { return read_callbacks_; }
```

Comment:
this doesn't have hacky dynamic casts so I think we could make it a regular accessor and remove the forTests label

Reply:
Done.